src/Mvc/Mvc.ApiExplorer/src/EndpointMetadataApiDescriptionProvider.cs

```diff
@@ -114,17 +114,27 @@ private ApiDescription CreateApiDescription(RouteEndpoint routeEndpoint, string
             var acceptsMetadata = routeEndpoint.Metadata.GetMetadata<IAcceptsMetadata>();
             if (acceptsMetadata is not null)
             {
-                var acceptsRequestType = acceptsMetadata.RequestType;
-                var isOptional = acceptsMetadata.IsOptional;
-                var parameterDescription = new ApiParameterDescription
+                // Do not add an implicit body parameter description if a parameter already
+                // exists for something derived from the body's content, such as a form file.
+                var alreadyHasAnyBodyParameter = apiDescription.ParameterDescriptions.Any(x =>
+                    x.Source == BindingSource.Body ||
+                    x.Source == BindingSource.Form ||
+                    x.Source == BindingSource.FormFile);
```

Comment:
Why didn't we need this check before and what makes it necessary now?

Reply:
I believe this is just something that wasn't catered for initially as with form files not being supported the code wouldn't get this far with something the parameter description would be wrong for. Without this check, it'll generate a parameter description that doesn't render appropriately in Swagger UI for dealing with a file.

I'll re-check this when I rebase and grab some screenshots of the before-after UI behaviour from rendering the OpenAPI document.

Reply:
Right, so this is because the code currently assumes that if the endpoint has an `IAcceptsMetadata`, it comes from the body and adds the parameter to the parameter descriptions because it's skipped here:

https://github.com/dotnet/aspnetcore/blob/a5a1e9a9d8631caf39366c11511c0cd126a02b34/src/Mvc/Mvc.ApiExplorer/src/EndpointMetadataApiDescriptionProvider.cs#L150-L155

This is because previously `IAcceptsMetadata` was only added for a JSON request body:

https://github.com/dotnet/aspnetcore/blob/a5a1e9a9d8631caf39366c11511c0cd126a02b34/src/Http/Http.Extensions/src/RequestDelegateFactory.cs#L892-L912

This new check prevents this "default" parameter being added for the body, because a more specific parameter for the `IFormFile` will have been added already. The route already has some `IAcceptsMetadata` in the case of form files as `IAcceptsMetadata` is added for them in these two places:

https://github.com/dotnet/aspnetcore/blob/05cbaa0f4bb99a8d4f54daba4fb7e85483d324aa/src/Http/Http.Extensions/src/RequestDelegateFactory.cs#L1044-L1060

https://github.com/dotnet/aspnetcore/blob/05cbaa0f4bb99a8d4f54daba4fb7e85483d324aa/src/Http/Http.Extensions/src/RequestDelegateFactory.cs#L1062-L1082

Alternatively I could remove this "default" behaviour for the body by removing the skip for `BindingSource.Body` and having it dealt with explicitly in `CreateApiParameterDescription()` instead.

Reply:
We want to support people manually consuming the request body and then specifying a type for the body "parameter" via `.Accepts<TRequest>(...)` or `[Consumes(...)]`. This is the reason for the current logic that adds an ApiParameterDescription for endpoint IAcceptsMetadata. As a consequence, we skip `source == BindingSource.Body` parameters in CreateApiParameterDescription() just to avoid duplicates.

> Alternatively I could remove this "default" behaviour for the body by removing the skip for BindingSource.Body and having it dealt with explicitly in CreateApiParameterDescription() instead.

I think this would probably the best approach if we could distinguish IAcceptsMetadata added by RequestDelegateFactory by default vs. IAcceptsMetadata added manually. Then we could only add ApiParameterDescriptions for manually-added IAcceptsMetadata. I'm not sure how we'd do that though.

The more consistent approach would be to also skip `source == BindingSource.FormFile`, but this loses info like the name of the form field(s) and whether there are multiple parameters for multiple fields.